core/src/unique.rs

```diff
@@ -160,23 +223,37 @@ impl<T> UniqueConnec<T> {
             old @ UniqueConnecInner::Full { .. } => {
                 // Keep the old value.
                 *inner = old;
-                return future::Either::B(until);
+                if pass_through {
+                    return future::Either::B(future::Either::A(until));
+                } else {
+                    return future::Either::B(future::Either::B(future::ok(())));
+                }
             },
         };
         drop(inner);
 
+        struct Cleaner<T>(Weak<Mutex<UniqueConnecInner<T>>>);
+        impl<T> Drop for Cleaner<T> {
+            #[inline]
+            fn drop(&mut self) {
+                if let Some(inner) = self.0.upgrade() {
+                    *inner.lock() = UniqueConnecInner::Empty;
+                }
+            }
+        }
+        let cleaner = Cleaner(Arc::downgrade(&self.inner));
+
         // The mutex is unlocked when we notify the pending tasks.
         for task in tasks_to_notify {
-            task.notify();
+            task.1.notify();
         }
 
-        let inner = self.inner.clone();
         let fut = until
             .select(on_clear_rx.then(|_| Ok(())))
             .map(|((), _)| ())
             .map_err(|(err, _)| err)
             .then(move |val| {
-                *inner.lock() = UniqueConnecInner::Empty;
+                drop(cleaner);      // Make sure that `cleaner` gets called there.
```

Comment:
Since `cleaner` is moved here,  would it not be sufficient to declare `cleaner = Arc::downgrade(&self.inner);` above and put the `Cleaner::drop` body in here? Why a custom `Drop` impl instead?

Reply:
In case the user drops the future.
`then` is only reached if you execute the future, but not if you just destroy it.